File: packages/realtime-protocol/src/mutation.ts
import { z } from "zod";
import {
  ClientUpdateIdSchema,
  EditSessionIdSchema,
  ReplicaIdSchema,
  U64Schema,
} from "./identity";
import { DEFAULT_REALTIME_LIMITS_V1 } from "./limits";
import { utf8Length } from "./unicode";

export const MUTATION_ORIGINS = [
  "human",
  "suggestion_accept",
  "version_restore",
  "external_small_save",
  "external_bulk_apply",
  "import",
] as const;

export type MutationOrigin = (typeof MUTATION_ORIGINS)[number];

const utf8String = (label: string, minimum: number, maximum: number) => z.string()
  .refine((value) => {
    let length: number;
    try {
      length = utf8Length(value, label);
    } catch {
      return false;
    }
    return length >= minimum && length <= maximum;
  }, `${label} must contain ${minimum} to ${maximum} UTF-8 bytes`);

export const AiAssistanceReceiptSchema = z.object({
  provider: utf8String("provider", 1, 128),
  model: utf8String("model", 1, 256),
  proposalIdentifier: utf8String("proposal identifier", 1, 256),
  acceptedDiff: utf8String("accepted diff", 0, 1_048_576),
}).strict();

export const MutationEnvelopeV1Schema = z.object({
  clientUpdateId: ClientUpdateIdSchema,
  replicaId: ReplicaIdSchema,
  clientSequence: U64Schema,
  editSessionId: EditSessionIdSchema,
  origin: z.enum(MUTATION_ORIGINS),
  assistance: AiAssistanceReceiptSchema.optional(),
  update: z.instanceof(Uint8Array).refine(
    (value) => value.length <= DEFAULT_REALTIME_LIMITS_V1.maxMutationUpdateBytes,
    "mutation update exceeds the default configured limit",
  ),
}).strict();

export type AiAssistanceReceipt = z.infer<typeof AiAssistanceReceiptSchema>;
export type MutationEnvelopeV1 = z.infer<typeof MutationEnvelopeV1Schema>;

File: crates/oleafly-realtime-protocol/src/mutation.rs
use crate::{ClientUpdateId, EditSessionId, ReplicaId};
use serde::{Deserialize, Serialize};

#[derive(Clone, Copy, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(rename_all = "snake_case")]
pub enum MutationOrigin {
    Human,
    SuggestionAccept,
    VersionRestore,
    ExternalSmallSave,
    ExternalBulkApply,
    Import,
}

#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct AiAssistanceReceipt {
    pub provider: String,
    pub model: String,
    pub proposal_identifier: String,
    pub accepted_diff: String,
}

impl AiAssistanceReceipt {
    pub fn validate(&self) -> crate::Result<()> {
        self.validate_with_limit(1_048_576)
    }

    pub fn validate_with_limit(&self, accepted_diff_limit: usize) -> crate::Result<()> {
        if self.provider.is_empty() || self.provider.len() > 128 {
            return Err(crate::ContractError::InvalidMutation(
                "assistance provider must contain 1 to 128 UTF-8 bytes".to_owned(),
            ));
        }
        if self.model.is_empty() || self.model.len() > 256 {
            return Err(crate::ContractError::InvalidMutation(
                "assistance model must contain 1 to 256 UTF-8 bytes".to_owned(),
            ));
        }
        if self.proposal_identifier.is_empty() || self.proposal_identifier.len() > 256 {
            return Err(crate::ContractError::InvalidMutation(
                "assistance proposal identifier must contain 1 to 256 UTF-8 bytes".to_owned(),
            ));
        }
        if self.accepted_diff.len() > accepted_diff_limit {
            return Err(crate::ContractError::InvalidMutation(
                "assistance accepted diff exceeds 1 MiB".to_owned(),
            ));
        }
        Ok(())
    }
}

#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct MutationEnvelopeV1 {
    pub client_update_id: ClientUpdateId,
    pub replica_id: ReplicaId,
    #[serde(with = "crate::canonical_u64")]
    pub client_sequence: u64,
    pub edit_session_id: EditSessionId,
    pub origin: MutationOrigin,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub assistance: Option<AiAssistanceReceipt>,
    pub update: Vec<u8>,
}
